src/editor/codelens.ts

```diff
@@ -0,0 +1,101 @@
+import {
+    CancellationToken,
+    CodeLens,
+    CodeLensProvider,
+    Event,
+    EventEmitter,
+    ExtensionContext,
+    Range,
+    TextDocument,
+    languages,
+    workspace
+} from 'vscode';
+import { ExtensionApi } from '../backend';
+import { AnalysisPathEvent, AnalysisPathKind } from '../backend/types';
+
+export class CodeLensStepsProvider implements CodeLensProvider {
+    constructor(ctx: ExtensionContext) {
+        ctx.subscriptions.push(this._onDidChangeCodeLenses = new EventEmitter());
+
+        ExtensionApi.diagnostics.diagnosticsUpdated(
+            this._onDidChangeCodeLenses.fire,
+            this._onDidChangeCodeLenses,
+            ctx.subscriptions
+        );
+
+        ctx.subscriptions.push(languages.registerCodeLensProvider('*', this));
+    }
+
+    private _onDidChangeCodeLenses: EventEmitter<void>;
+    public get onDidChangeCodeLenses(): Event<void> {
+        return this._onDidChangeCodeLenses.event;
+    }
+
+    provideCodeLenses(document: TextDocument, _token: CancellationToken): CodeLens[] {
+        if (ExtensionApi.diagnostics.selectedEntry === undefined ||
+            !workspace.getConfiguration('codechecker.editor').get('enableCodeLens')) {
+            return [];
+        }
+
+        const entry = ExtensionApi.diagnostics.selectedEntry;
+        const fullPath = entry.diagnostic.path
+            .filter(elem => elem.kind === AnalysisPathKind.Event) as AnalysisPathEvent[];
+
+        if (fullPath.length === 0) {
+            return [];
+        }
+
+        const codeLenses = [];
+
+        for (const [idx, pathItem] of fullPath.entries()) {
+            if (entry.diagnostic.files[pathItem.location.file] !== document.uri.fsPath) {
+                continue;
+            }
+
+            // TODO: Handle multiple ranges
+            const range = new Range(
+                pathItem.location.line-1,
+                pathItem.location.col-1,
+                pathItem.location.line-1,
+                pathItem.location.col,
+            );
+
+            // FIXME: Handle multiple steps in the same line
```

Comment:
For me it shows the following codelens if there are multiple events for the same line:
![image](https://user-images.githubusercontent.com/6695818/141760605-0d9e0eac-a6cc-459a-bc3d-448c8fde3b35.png)

Does CodeLens support adding multiline notes?

Reply:
There's no support for multiline notes, as far as I know.

I'm not sure what the best way to display this would be.

Maybe something like this, if they're consecutive?
``5. Description 1 | 6. Description 2 | Previous step | Next step``
(prev. links to 4., and next links to 7.)

If they're not consecutive, I can show the previous/next step for each one like above, but in the proper order.

Reply:
The steps are now ordered, and also don't show extra step buttons.

Reply:
I like this change, thank you. Do we still need this FIXME note?

Reply:
Most definitely not! Removed, thanks.